Edits to Logging Guide (#10365)
keycloak/keycloak: docs/guides/src/main/server/logging.adoc

summary="Learn how to configure Logging"
includedOptions="log-*">

Keycloak uses the jboss logmanager logging framework. The following is a high-level overview for the available log handlers:

> 

Reviewer: I'd suggest rewriting to "Keycloak uses the JBoss Logging framework...."

Author: Thanks, @abstractj. Corrected.

* root
** console (_default_)
** file

> == Logging: Root configuration

Author: @DGuhr I am wondering if "Root" should be removed  because this section also includes category-specific logging configuration.

Reviewer 2: Makes sense. sry 4 late response here, too.

Logging is done on a per-category basis in Keycloak. You can configure logging for the root log level or for more specific categories such as `org.hibernate` or `org.keycloak`. This guide describes how to configure logging.

> === Root log levels

Author: @DGuhr I am wondering if "Root" should be removed  because this section also applies to log levels for category-specific logging unless I am mistaken.

Reviewer 2: Yeh, makes sense. and sorry for the late response :)

|WARN|A non-critical error or problem that might not require immediate correction.
|INFO|Keycloak lifecycle events or important information. Low frequency.
|DEBUG|More detailed information for debugging purposes, such as database logs. Higher frequency.
|TRACE|Most detailed debugging information. Very high frequency.
|ALL|Special level for all log messages.
|OFF|Special level to turn logging off entirely (not recommended).
|====

=== Configuring the root log level 
When no log level configuration exists for a more specific category logger, the enclosing category is used instead. When there is no enclosing category, the root logger level is used.

To set the root log level, enter the following command:

<@kc.start parameters="--log-level=<root-level>"/>

Use these guidelines for this command:

* For `_<root-level>_`, supply a level defined in the preceding table.
* The log level is case-insensitive. For example, you could either use `DEBUG` or `debug`.
* If you were to accidentally set the log level twice, the last occurrence in the list becomes the log level. For example, if you included the syntax `--log-level=info,...,DEBUG,...`, the root logger would be `DEBUG`.

> == Configuring category-specific log levels

Reviewer 2: when removing "root" from the headlines, this should be `===` again i think.

.Example
<@kc.start parameters="--log-level=INFO,org.hibernate:debug,org.hibernate.hql.internal.ast:info"/>

This example sets the following log levels:

* Root log level for all loggers is set to INFO.
* The hibernate log level in general is set to debug.
* To keep SQL abstract syntax trees from creating verbose log output, the specific subcategory `org.hibernate.hql.internal.ast` is set to info. As a result, the SQL abstract syntax trees are omitted instead of appearing at the `debug` level.

== Logging format configuration
Keycloak uses a pattern-based logging formatter that generates human-readable text logs. The logging format template for these lines can be applied at the root level. The default format template is:

* `%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c] (%t) %s%e%n`

The format string supports the symbols in the following table:

|====
|Symbol|Summary|Description
|%%|%|Renders a simple % character.
|%c|Category|Renders the log category name.
|%d{xxx}|Date|Renders a date with the given date format string.String syntax defined by `java.text.SimpleDateFormat`

> |%e|Exception|Renders any thrown exception that exists.

Reviewer 2: ```suggestion
|%e|Exception|Renders a thrown exception
```

Reviewer 2: It does not necessarily throw every exception, but only the explicit exception thrown, so perhaps just stick with the suggestion here?

|%h|Hostname|Renders the simple host name.
|%H|Qualified host name|Renders the fully qualified hostname, which may be the same as the simple host name, depending on the OS configuration.
|%i|Process ID|Renders the current process PID.

> |%m|Full Message|Renders the log message plus any exception that exists.

Reviewer 2: ```suggestion
|%m|Full Message|Renders the log message and an exception, if thrown.
```

Reviewer 2: see above. "any that exists" is not true. One line of log contains one exception. There may be more, e.g. when you set log to trace there are actually exceptions right now. These would not be printed, but "any...that exists" suggests so imo.

|%p|Level|Renders the log level of the message.
|%r|Relative time|Render the time in milliseconds since the start of the application log.
|%s|Simple message|Renders only the log message without exception trace.
|%t|Thread name|Renders the thread name.
|%t{id}|Thread ID|Render the thread ID.
|%z{<zone name>}|Time zone|Sets the time zone of log output to <zone name>.
|====

=== Setting the logging format
To set the logging format for a logged line, perform these steps:

. Build your desired format template using the table above
. Enter the following command:
+
<@kc.start parameters="--log-format=\"\'<format>\'\""/>

Note that you need to escape characters when invoking commands containing special shell characters such as `;` using the CLI. Therefore, consider setting it in the configuration file instead.

.Example: Abbreviate the fully qualified category name

<@kc.start parameters="\"\'%d{yyyy-MM-dd HH:mm:ss,SSS} %-5p [%c{3.}] (%t) %s%e%n\'\""/>

The example above abbreviates the category name, which could get rather long in some cases, to three characters by setting `[%c{3.}]` in the template instead of the default `[%c]`.

== Raw quarkus logging properties

> At publication time for this guide, the logging features of the Quarkus-based Keycloak are basic, yet powerful. Nevertheless, expect more to come and feel free to join the https://github.com/keycloak/keycloak/discussions/8870[discussion] at GitHub.

Reviewer 2: ```suggestion
At publication time for this guide, the logging features of the Quarkus-based Keycloak are basic, yet powerful. Nevertheless, expect more to come and feel free to join the https://github.com/keycloak/keycloak/discussions/categories/keycloak-x-quarkus-distribution[discussion] at GitHub.
```

Reviewer 2: The mentioned discussion is closed now, for successors of parts to come. I think we go better with a more generic link here to not have to update it over and over ;)

> If you need a temporary solution, such as for logging to a file or using syslog instead of the console, you can check out the https://github.com/keycloak/keycloak/discussions/8870[Quarkus logging guide]. You might be able to use all properties mentioned there as long as no other than the base logging dependency is involved. For example, you might be able to set the log handler to file, but not to use JSON output yet. You would need to provide another dependency for JSON output to work.

Reviewer 2: ```suggestion
If you need a temporary solution, such as for logging to a file or using syslog instead of the console, you can check out the https://quarkus.io/guides/logging[Quarkus logging guide]. You are able to use all properties mentioned in the guide as long as no other than the base logging dependency is involved. For example, you are able to set the log handler to file or syslog, but not to use JSON output yet. You would need to provide another dependency for JSON output to work.
```

Reviewer 2: 1) Link was wrong, oops. changed it.
2) "might be" implies to me a non-existing insecurity if the described process is possible. "it might be so, it might not."-  It is possible, so just use "are" instead?

To use unstructured logging, enter the following command:

<@kc.start parameters="--log-console-output=default"/>

.Example Log Message:
[source, bash]
----
2022-03-02 10:36:50,603 INFO  [io.quarkus] (main) Keycloak 18.0.0-SNAPSHOT on JVM (powered by Quarkus 2.7.2.Final) started in 3.615s. Listening on: http://0.0.0.0:8080
----

=== Colors
Colored console log output for unstructured logs is disabled by default. Colors may improve readability, but they can cause problems when shipping logs to external log aggregation systems. If you want to enable or disable color-coded console log output, enter following command:

<@kc.start parameters="--log-console-color=<false|true>"/>

== File logging
As an alternative to logging to the console, you can use unstructured logging to a file.

=== Enable file logging
Logging to a file is disabled by default. To enable it, enter the following command:

<@kc.start parameters="--log=console,file"/>

A log file named `keycloak.log` is created inside the `data/log` directory of your Keycloak installation.

=== Configuring the location and name of the log file

To change where the log file is created and the file name, perform these steps:

> . Find a writable directory to store the log file.

Reviewer 2: "Find" sounds a bit weird to me. "Use" or "Create" are better IMO.

+

> If the directory is not writable,  Keycloak will start correctly, but no log file is created.

Reviewer 2: ```suggestion
If the directory is not writable,  Keycloak will start correctly, but no log file is created and an error is shown.
```

. Enter this command:
+
<@kc.start parameters="--log=console,file --log-file=<path-to>/<your-file.log>"/>

=== Configuring the file handler format
To configure a different logging format for the file log handler, enter the following command:

<@kc.start parameters="--log-file-format=<pattern>"/>

For details on the available pattern configuration, see <<Configuring the console log format>>.

> == Raw quarkus logging properties

Reviewer 2: Pls remove this whole section. we are including json logging and file logging now, and i don't think we should advertise using raw properties anymore but instead let them create issues on our side.

+
If the directory is not writable, Keycloak will start correctly, but it will issue an error and no log file will be created.

. Enter this command:
+
<@kc.start parameters="--log=console,file --log-file=<path-to>/<your-file.log>"/>

=== Configuring the file handler format
To configure a different logging format for the file log handler, enter the following command:

<@kc.start parameters="--log-file-format=<pattern>"/>

Please see the <<Configuring the console log format>> section in this guide for more information and a table of the available pattern configuration.

== Centralized logging using Gelf
Keycloak can send logs to a centralized log management system such as the following:

* Graylog
* Logstash, inside the Elasticsearch, Logstash, Kibana (ELK) logging stack
* Fluentd, inside the Elasticsearch, Fluentd, Kibana (EFK) logging stack

Keycloak uses the https://quarkus.io/guides/centralized-log-management[Quarkus Logging Gelf] extension to support these environments.

=== Enabling the Gelf handler
To enable logging using Gelf, add it to the list of activated log handlers.

.Example:
<@kc.start parameters="--log=console,gelf"/>

> === Configuring the Gelf handler

Reviewer 2: I changed this to use GELF on purpose as part of #13422, looking at the actual https://docs.graylog.org/docs/gelf docs and others. It is an acronym for Graylog Extended Log Format. Pls change back to `GELF` everywhere, so we are consistent with other original sources and our CLI.

<@kc.start parameters="--log=console,gelf --log-gelf-host=myhost --log-gelf-port=12345"/>

When the Gelf handler is enabled, the host is using `localhost` as host value and UDP for communication. To use TCP instead of UDP, prefix the host value with `tcp:`. The Default port is `12201`.

.Include or exclude Stacktraces
Keycloak includes the complete Stacktrace inside the `StackTrace` field. To exclude this field, enter the following command:

<@kc.start parameters="--log=console,gelf --log-gelf-include-stack-trace=false"/>

.Configure the timestamp format
You can change the format of the `timestamp` field. For example, you can include the date and time down to seconds by entering the following command:

<@kc.start parameters="--log=console,gelf --log-gelf-timestamp-format=\"\'yyyy-MM-dd HH:mm:ss\'\""/>

Alternatively, you could use the config file to avoid escaping:

[source, conf]
----
log-gelf-timestamp-format=yyyy-MM-dd HH:mm:ss
----

The default timestamp format is `yyyy-MM-dd HH:mm:ss,SSS`. You can use the https://docs.oracle.com/javase/10/docs/api/java/text/SimpleDateFormat.html[available SimpleDateFormat patterns] to define an appropriate timestamp.

.Configure the facility

> The `facility` field is an indicator of the process or program that is the source of log messages. The default value is keycloak. To set this field to your preferred identifier, enter the following command:

Reviewer 2: ```suggestion
The `facility` field is an indicator of the process or program that is the source of log messages. The default value is `keycloak`. To set this field to your preferred identifier, enter the following command:
```
